execution/execution.c

```diff
@@ -0,0 +1,219 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   execution.c                                        :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: asyed <asyed@student.42.fr>                +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2018/03/21 21:16:12 by asyed             #+#    #+#             */
+/*   Updated: 2018/04/04 17:34:58 by asyed            ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "exec.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <errno.h>
+
+struct s_ophandlers	op_handlers[] = {
+	{&op_pipe_check, &op_pipe_exec},
+	{&op_or_check, &op_or_exec},
+	{&op_and_check, &op_and_exec},
+	{NULL, NULL},
+};
+
+int		run_operation(t_ast *curr, uint8_t wait)
+{
+	pid_t	pid;
+	int 	res;
+
+	if (!curr || *(curr->type) == operator)
+		return (EXIT_FAILURE);
+	if ((res = builtin_handler(curr)) != -1)
+		return (res);
+	if ((pid = fork()) == -1)
+		return (EXIT_FAILURE);
+	if (pid == 0)
+	{
+		if (handle_redirection(curr))
+			exit(EXIT_FAILURE);
```

Comment:
Is it possible to return instead of exiting ? (Gaetan requirement) If it is not it is ok. Maybe it has to do with the issue when a command hit a segfault

Reply:
Its exiting within the forked process itself so that it doesn't daisy chain to continue to process anything unnecessarily (that way waiting still works and once its done evaluating its completely done with the forked process)